Math/src/main/java/org/tribuo/math/la/DenseMatrix.java

```diff
@@ -831,4 +1371,405 @@ public MatrixTuple next() {
+
+            // Solve matrix^T . output = y
+            for (int i = lMatrix.dim1-1; i >= 0; i--) {
+                double sum = output[i];
+                for (int j = i+1; j < lMatrix.dim1; j++) {
+                    sum -= lMatrix.values[j][i] * output[j];
+                }
+                output[i] = sum / lMatrix.values[i][i];
+            }
+
+            return new DenseVector(output);
+        }
+
+        /**
+         * Solves the system A * X = B, where B is the input matrix, and A is the matrix which
+         * produced this Cholesky factorization.
+         * @param matrix The input matrix B.
+         * @return The matrix X.
+         */
+        public DenseMatrix solve(Matrix matrix) {
+            if (matrix.getDimension1Size() != lMatrix.dim1) {
+                throw new IllegalArgumentException("Size mismatch, expected " + lMatrix.dim1 + ", received " + matrix.getDimension1Size());
+            }
+            final int outputDim1 = lMatrix.dim1;
+            final int outputDim2 = matrix.getDimension2Size();
+            final DenseMatrix output = new DenseMatrix(matrix);
+            final double[][] outputArr = output.values;
+
+            // Solve L.Y = B
+            for (int i = 0; i < outputDim1; i++) {
+                for (int j = 0; j < outputDim2; j++) {
+                    for (int k = 0; k < i; k++) {
+                        outputArr[i][j] -= outputArr[k][j] * lMatrix.values[i][k];
+                    }
+                    // scale by diagonal
+                    outputArr[i][j] /= lMatrix.values[i][i];
+                }
+            }
+
+            // Solve L^T.X = Y
+            for (int i = outputDim1 - 1; i >= 0; i--) {
+                for (int j = 0; j < outputDim2; j++) {
+                    for (int k = i + 1; k < outputDim2; k++) {
+                        outputArr[i][j] -= outputArr[k][j] * lMatrix.values[k][i];
+                    }
+                    // scale by diagonal
+                    outputArr[i][j] /= lMatrix.values[i][i];
+                }
+            }
+
+            return output;
+        }
+
+        /**
+         * Generates the inverse of the matrix with this Cholesky factorization.
+         * @return The matrix inverse.
+         */
+        public DenseMatrix inverse() {
+            return solve(DenseSparseMatrix.createIdentity(lMatrix.dim1));
+        }
+    }
+
+    /**
+     * The output of a successful LU factorization.
+     * <p>
+     * Essentially wraps a pair of {@link DenseMatrix}, but has additional
+     * operations which allow more efficient implementations when the
+     * matrices are known to be the result of a LU factorization.
+     * <p>
+     * Mutating the wrapped matrices will cause undefined behaviour in the methods
+     * of this class.
+     */
+    public static final class LUFactorization {
+        /**
+         * The lower triangular matrix, with ones on the diagonal.
+         */
+        public final DenseMatrix l;
```

Comment:
this member name is unfortunate as it is hard to read is at an l or an I or 1?  Perhaps 'lower' and 'upper' would be better here.

Reply:
Sure.